### PR TITLE
TreeKEM

### PR DESCRIPTION
**Pull request by beurdouche**
Work in progress...

### CODE REVIEW

**Comment by beurdouche** [comment]
Still WIP

File: draft-barnes-mls-protocol.md [hunk 1]
```diff
@@ -126,8 +126,10 @@ size of the group.
 
 In this document, we describe a protocol based on tree structures
 that enable asynchronous group keying with forward secrecy and
-post-compromise security.  The use of "asynchronous ratcheting
-trees" {{art}} allows the members of the group to derive and update
+post-compromise security.  This document propose two approaches, one
```
Comment (ekr) [inline draft-barnes-mls-protocol.md]: s/propose/describes two candidate approaches/

File: draft-barnes-mls-protocol.md [hunk 2]
```diff
@@ -500,30 +502,41 @@ A   B*    C    D
 
 ## Ratchet Trees
 
-Ratchet trees are used for generating shared group secrets. These are
-constructed as a series of Diffie-Hellman keys in a binary tree arrangement,
-with each user knowing their direct path, and thus being able to compute the
-shared root secret.
+This document present two different potential approaches, ART and TreeKEM,
+for generating and managing group secrets using Ratchet tree datastructures.
+These are constructed as a series of asymetric keypairs in a left-balanced
```
Comment (ekr) [inline draft-barnes-mls-protocol.md]: s/Those/Both/

File: draft-barnes-mls-protocol.md [hunk 3]
```diff
@@ -500,30 +502,41 @@ A   B*    C    D
...
-* private, public = Derive-Key-Pair(secret)
+Ratchet trees constructed in ART or TreeKEM provide the property that one must hold at
+least one private key from the tree to compute the secret root key. With all
+participants holding one leaf private key; this allows any individual to update
+their own key and change the shared root key, such that only group members can
+compute the new key.
+
+### Ratchet Trees for ART
+
+In ART the value of the parent is computed as follows:
```
Comment (ekr) [inline draft-barnes-mls-protocol.md]: parent secret

File: draft-barnes-mls-protocol.md [hunk 4]
```diff
@@ -541,11 +554,25 @@ DH(DH(AB), DH(CD))
...
 ~~~~~
 
-Ratchet trees constructed this way provide the property that one must hold at
-least one private key from the tree to compute the secret root key. With all
-participants holding one leaf private key; this allows any individual to update
-their own key and change the shared root key, such that only group members can
-compute the new key.
+### Ratchet Trees for TreeKEM
+
+The key computations for each Node differs in TreeKEM. The general idea behind
```
Comment (ekr) [inline draft-barnes-mls-protocol.md]: computations differ

File: draft-barnes-mls-protocol.md [hunk 5]
```diff
@@ -541,11 +554,25 @@ DH(DH(AB), DH(CD))
...
+TreeKEM is to obtain better messaging complexity on concurrent events by not
+forcing contributivity of all leaves to the keys of the *intermediate* nodes.
+Instead, modification of a key will lead to sending to the copath the
+information necessary to be able to recompute the root key through a
+Key Encapsulation Mechanism (KEM) which can be typically post-quantum resistant.
```
Comment (ekr) [inline draft-barnes-mls-protocol.md]: I would say:

```
improve the efficiency of concurrent updates by allowing them to be merged, rather
than forcing one update to be selected and others to be discarded. Instead, the 
last update to the key at a given node replaces all previous updates.
```

File: draft-barnes-mls-protocol.md [hunk 6]
```diff
@@ -541,11 +554,25 @@ DH(DH(AB), DH(CD))
...
+
+* parent_secret = Hash(A OR B)
+* parent_private, parent_public = Derive-Key-Pair(parent_secret)
+
+~~~~~
+   H(H(A||B)||H(C||D))
+       /      \
+   H(A||B)   H(C||D)
```
Comment (ekr) [inline draft-barnes-mls-protocol.md]: Don't use the syntax ||. That means concatenation in a lot of IETF specs.  maybe OR.

File: draft-barnes-mls-protocol.md [hunk 7]
```diff
@@ -541,11 +554,25 @@ DH(DH(AB), DH(CD))
...
+    /  \      /  \
+   A    B    C    D
```
Comment (ekr) [inline draft-barnes-mls-protocol.md]: This is kind of confusing. I think you need to start small. Why don't you crib my text from the mail to the listl.

File: draft-barnes-mls-protocol.md [hunk 8]
```diff
@@ -692,6 +720,19 @@ infinity (O), (2) verify that for Y = (x, y) both integers are in the correct
 interval, (3) ensure that (x, y) is a correct solution to the elliptic curve equation.
 For these curves, implementers do not need to verify membership in the correct subgroup.
 
+### TreeKEM: ECIES or any PQ KEM with SHA-256
+
+This ciphersuite uses the following primitives:
+
+* Hash function: SHA-256
+* Key Encapsulation: ECIES or any PQ KEM
```
Comment (ekr) [inline draft-barnes-mls-protocol.md]: I think you can remove PQ KEM

File: draft-barnes-mls-protocol.md [hunk 9]
```diff
@@ -692,6 +720,19 @@ infinity (O), (2) verify that for Y = (x, y) both integers are in the correct
...
+
+A significant advantage of TreeKEM over ART is that intermediate node secrets
+are sent via a Key Encapsulation Mechanism (KEM) which can be cryptographically
+resistant to quantic adversaries unlike Elliptic Curve operations.
```
Comment (ekr) [inline draft-barnes-mls-protocol.md]: Eh, I wouldn't bother with this. Also, quantic isn't idiomatic here.

File: draft-barnes-mls-protocol.md [hunk 10]
```diff
@@ -850,7 +891,8 @@ In MLS, these changes are accomplished by broadcasting "handshake"
 messages to the group.  Note that unlike TLS and DTLS, there is not
 a consolidated handshake phase to the protocol.  Rather, handshake
 messages are exchanged throughout the lifetime of a group, whenever
-a change is made to the group state.
+a change is made to the group state. This means a an unbounded number
```
Comment (ekr) [inline draft-barnes-mls-protocol.md]: "a an".

File: draft-barnes-mls-protocol.md [hunk 11]
```diff
@@ -1187,6 +1276,7 @@ The requirement for this is that all participants know these values.
 If additional clear-text fields are attached to messages (like the counter), those
 fields MUST be protected by a signed message envelope.
 
+[BB. This seems both odd and unwise...]
```
Comment (ekr) [inline draft-barnes-mls-protocol.md]: Remove.

File: draft-barnes-mls-protocol.md [hunk 12]
```diff
@@ -1051,34 +1081,84 @@ root node of the ratchet tree.
...
+~~~~~
+
+These messages are sent to each individual nodes, in the copath
+of a modified leaf, encrypted under their public key as follows:
+
+~~~~~
+struct {
+    uint32 length;
+    opaque content[KEMCiphertext.length];
+} KEMCiphertext;
```
Comment (ekr) [inline draft-barnes-mls-protocol.md:1329]: This isn't right. You send to the highest joint parent.
-> Ben: no, it doesn't work. You can have insider attacks if you do that...

**Review by ekr — COMMENTED** [review]
You really need to explain the "last one in wins" property of TreeKEM. Also, about how you have to age out the existing keys, and process the new ones.

File: draft-barnes-mls-protocol.md [hunk 13]
```diff
@@ -556,20 +556,40 @@ A    B    C    D
 
 ### Ratchet Trees for TreeKEM
 
-The key computations for each Node differs in TreeKEM. The general idea behind
+The key computation differ for each Node in TreeKEM. The general idea behind
```
Comment (ekr) [inline draft-barnes-mls-protocol.md]: computations differ.

File: draft-barnes-mls-protocol.md [hunk 14]
```diff
@@ -582,16 +602,17 @@ leaves when participants are deleted from the group.
...
-  DH(A, DH(CD))
+  COMPUTE(A, COMPUTE(CD))
    /      \
-  A       DH(CD)
+  A       COMPUTE(CD)
  / \      /  \
 A   _    C    D
 ~~~~~
 
+The COMPUTE operation represents any mechanism provider by ART or TreeKEM.
```
Comment (ekr) [inline draft-barnes-mls-protocol.md]: The problem with COMPUTE here is that it's not a function of CD in TreeKEM

File: draft-barnes-mls-protocol.md [hunk 15]
```diff
@@ -594,6 +594,55 @@ be kept in the new root node.
...
+In order to do so, the participant will compute the sequence of hashes
+H(C'), H(H(C')) up-to the root node. It will also compute the sequence
+of ciphertexts, containing for each group on the copath to the root the
+new key for its parent. In this example, participant D would receive the
+secret H(C') for its parent by C. The node on AB being in the copath of
+C to the root, it would (hence A and B would to) receive the secret for
+its parent H(H(C')) encrypted under its current private key.
+
+Note that encrypting the new key to the previous root private key to
+distribute it to all participants is unsecure because it leads to insider
```
Comment (ekr) [inline draft-barnes-mls-protocol.md]: insecure,

File: draft-barnes-mls-protocol.md [hunk 16]
```diff
@@ -594,6 +594,55 @@ be kept in the new root node.
    A    B     C'   D
 ~~~~~
 
+This behavior allows completely concurrent updates which is likely to
+be most of the cases of concurrency related issues in during the lifetime
+of the group.
```
Comment (ekr) [inline draft-barnes-mls-protocol.md]: This sentence is awkeward.

File: draft-barnes-mls-protocol.md [hunk 17]
```diff
@@ -594,6 +594,55 @@ be kept in the new root node.
...
+that will follow.
+
+While operations leading to key update may be completely concurrent, it
+is not the case that all operations can permute. Indeed, a concurrent
+execution of an Update with a Delete operation may lead a user to be locked
+out by a delete and then reinstalled in the group by receiving an update.
+To avoid this issue, a participant receiving two concurrent messages which
+are not Update MUST process all Update messages before performing the
+operations for the Delete. Fortunately in MLS the Delivery Service can
+enforce total ordering of the messages to avoid this problem all together.
```
Comment (ekr) [inline draft-barnes-mls-protocol.md:747]: This still doesn't cover how long you need to hold onto the previous keys.

File: draft-barnes-mls-protocol.md [hunk 18]
```diff
@@ -594,6 +594,55 @@ be kept in the new root node.
...
+
+MLS requires that all operations performing changes to a group state
+MUST be authenticated. Having access to group root secret means that
+the participant is a legitimate receiver of the messages but does not
+mean that participants can't act on behalf of each other. To protect
```
Comment (ekr) [inline draft-barnes-mls-protocol.md]: can't act? Huh?

File: draft-barnes-mls-protocol.md [hunk 19]
```diff
@@ -594,6 +594,55 @@ be kept in the new root node.
...
+against this, the participant MUST sign the Handshake messages as those
+are the ones triggering group operations. On receiption of a Handshake
```
Comment (ekr) [inline draft-barnes-mls-protocol.md]: reception.

File: draft-barnes-mls-protocol.md [hunk 20]
```diff
@@ -594,6 +594,55 @@ be kept in the new root node.
...
+message, a participant MUST verify the signature before performing the
+group operation on its local state. This avoids, for example, a
+participant to claim that another participant has change its key.
```
Comment (ekr) [inline draft-barnes-mls-protocol.md]: Has changed?

**Comment by ekr** [comment]
This is starting to need some big-time copy editing.

**Review by ekr — CHANGES_REQUESTED** [review]

File: draft-barnes-mls-protocol.md (hunk 17 above)
Comment (beurdouche) [reply draft-barnes-mls-protocol.md:747]: I picked 7 days at random, but we need more user data to have something reasonable.

**Pull request merged**